File: dart/lib/src/trie/aho_corasick.dart
import '../normalization/normalized_text.dart';

/// One raw hit from a scan.
class RawHit {
  const RawHit(this.patternId, this.end);

  /// The pattern that matched.
  final int patternId;

  /// The index in the scanned text of the pattern's last character.
  final int end;
}

/// A prefix trie with failure links.
///
/// It finds every pattern in one pass over the text, so the cost stays linear
/// in the length of the text.
class AhoCorasick {
  AhoCorasick() {
    _children.add(<int, int>{});
    _terminal.add(-1);
  }

  final List<Map<int, int>> _children = <Map<int, int>>[];
  final List<int> _lengths = <int>[];
  final List<int> _terminal = <int>[];
  late List<int> _fail;
  late List<int> _outputLink;
  bool _built = false;

  /// How many patterns the trie holds.
  int get patternCount => _lengths.length;

  /// The character count of one pattern.
  int lengthOf(int patternId) => _lengths[patternId];

  /// Inserts one pattern.
  ///
  /// Returns the id of the pattern. A pattern that is already present keeps its
  /// first id, so the trie never holds a duplicate.
  int add(List<int> pattern) {
    if (_built) {
      throw StateError('Add a pattern before you call build.');
    }
    if (pattern.isEmpty) {
      throw ArgumentError.value(
          pattern, 'pattern', 'A pattern must hold at least one character.');
    }

    int node = 0;
    for (final int c in pattern) {
      int? next = _children[node][c];
      if (next == null) {
        next = _children.length;
        _children.add(<int, int>{});
        _terminal.add(-1);
        _children[node][c] = next;
      }
      node = next;
    }

    if (_terminal[node] >= 0) {
      return _terminal[node];
    }

    final int id = _lengths.length;
    _lengths.add(pattern.length);
    _terminal[node] = id;
    return id;
  }

  /// Computes the failure links. Call this once, after the last [add].
  void build() {
    if (_built) {
      return;
    }

    final int count = _children.length;
    _fail = List<int>.filled(count, 0);
    _outputLink = List<int>.filled(count, -1);

    final List<int> queue = <int>[];
    int head = 0;

    // Depth 1 always fails back to the root.
    for (final int child in _children[0].values) {
      _fail[child] = 0;
      queue.add(child);
    }

    while (head < queue.length) {
      final int node = queue[head++];

      final int failNode = _fail[node];
      _outputLink[node] =
          _terminal[failNode] >= 0 ? failNode : _outputLink[failNode];

      _children[node].forEach((int character, int child) {
        int candidate = _fail[node];
        while (candidate != 0 && !_children[candidate].containsKey(character)) {
          candidate = _fail[candidate];
        }

        final int? target = _children[candidate][character];
        _fail[child] = (target != null && target != child) ? target : 0;
        queue.add(child);
      });
    }

    _built = true;
  }

  /// Finds every pattern in the text and appends each hit to [hits].
  void scan(NormalizedText text, List<RawHit> hits) {
    if (!_built) {
      throw StateError('Call build before you scan.');
    }
    if (_lengths.isEmpty) {
      return;
    }

    int node = 0;
    final List<int> chars = text.chars;

    for (int i = 0; i < chars.length; i++) {
      final int character = chars[i];

      while (node != 0 && !_children[node].containsKey(character)) {
        node = _fail[node];
      }

      node = _children[node][character] ?? 0;

      int output = _terminal[node] >= 0 ? node : _outputLink[node];
      while (output >= 0) {
        hits.add(RawHit(_terminal[output], i));
        output = _outputLink[output];
      }
    }
  }
}

File: dotnet/src/Vulgarity/Trie/AhoCorasick.cs
using System;
using System.Collections.Generic;
using Vulgarity.Normalization;

namespace Vulgarity.Trie
{
    /// <summary>One raw hit from a scan.</summary>
    internal struct RawHit
    {
        /// <summary>The pattern that matched.</summary>
        public int PatternId;

        /// <summary>The index in the scanned text of the pattern's last character.</summary>
        public int End;

        public RawHit(int patternId, int end)
        {
            PatternId = patternId;
            End = end;
        }
    }

    /// <summary>
    /// A prefix trie with failure links. It finds every pattern in one pass over
    /// the text, so the cost stays linear in the length of the text.
    /// </summary>
    internal sealed class AhoCorasick
    {
        private readonly List<Dictionary<int, int>> _children = new List<Dictionary<int, int>>();
        private readonly List<int> _lengths = new List<int>();
        private readonly List<int> _terminal = new List<int>();
        private int[] _fail;
        private int[] _outputLink;
        private bool _built;

        public AhoCorasick()
        {
            _children.Add(new Dictionary<int, int>());
            _terminal.Add(-1);
        }

        /// <summary>How many patterns the trie holds.</summary>
        public int PatternCount
        {
            get { return _lengths.Count; }
        }

        /// <summary>The character count of one pattern.</summary>
        public int LengthOf(int patternId)
        {
            return _lengths[patternId];
        }

        /// <summary>Inserts one pattern.</summary>
        /// <returns>
        /// The id of the pattern. A pattern that is already present keeps its
        /// first id, so the trie never holds a duplicate.
        /// </returns>
        public int Add(int[] pattern)
        {
            if (_built)
            {
                throw new InvalidOperationException("Add a pattern before you call Build.");
            }

            if (pattern.Length == 0)
            {
                throw new ArgumentException("A pattern must hold at least one character.", "pattern");
            }

            int node = 0;
            for (int i = 0; i < pattern.Length; i++)
            {
                int next;
                if (!_children[node].TryGetValue(pattern[i], out next))
                {
                    next = _children.Count;
                    _children.Add(new Dictionary<int, int>());
                    _terminal.Add(-1);
                    _children[node][pattern[i]] = next;
                }

                node = next;
            }

            if (_terminal[node] >= 0)
            {
                return _terminal[node];
            }

            int id = _lengths.Count;
            _lengths.Add(pattern.Length);
            _terminal[node] = id;
            return id;
        }

        /// <summary>Computes the failure links. Call this once, after the last Add.</summary>
        public void Build()
        {
            if (_built)
            {
                return;
            }

            int count = _children.Count;
            _fail = new int[count];
            _outputLink = new int[count];
            for (int i = 0; i < count; i++)
            {
                _outputLink[i] = -1;
            }

            Queue<int> queue = new Queue<int>();

            // Depth 1 always fails back to the root.
            foreach (KeyValuePair<int, int> edge in _children[0])
            {
                _fail[edge.Value] = 0;
                queue.Enqueue(edge.Value);
            }

            while (queue.Count > 0)
            {
                int node = queue.Dequeue();

                int failNode = _fail[node];
                _outputLink[node] = _terminal[failNode] >= 0 ? failNode : _outputLink[failNode];

                foreach (KeyValuePair<int, int> edge in _children[node])
                {
                    int character = edge.Key;
                    int child = edge.Value;

                    int candidate = _fail[node];
                    int target;
                    while (candidate != 0 && !_children[candidate].ContainsKey(character))
                    {
                        candidate = _fail[candidate];
                    }

                    if (_children[candidate].TryGetValue(character, out target) && target != child)
                    {
                        _fail[child] = target;
                    }
                    else
                    {
                        _fail[child] = 0;
                    }

                    queue.Enqueue(child);
                }
            }

            _built = true;
        }

        /// <summary>Finds every pattern in the text and appends each hit to <paramref name="hits"/>.</summary>
        public void Scan(NormalizedText text, List<RawHit> hits)
        {
            if (!_built)
            {
                throw new InvalidOperationException("Call Build before you scan.");
            }

            if (_lengths.Count == 0)
            {
                return;
            }

            int node = 0;
            int[] chars = text.Chars;

            for (int i = 0; i < chars.Length; i++)
            {
                int character = chars[i];

                while (node != 0 && !_children[node].ContainsKey(character))
                {
                    node = _fail[node];
                }

                int next;
                node = _children[node].TryGetValue(character, out next) ? next : 0;

                int output = _terminal[node] >= 0 ? node : _outputLink[node];
                while (output >= 0)
                {
                    hits.Add(new RawHit(_terminal[output], i));
                    output = _outputLink[output];
                }
            }
        }
    }
}
